msbuild/Xamarin.MacDev.Tasks.Core/Tasks/CodesignTaskBase.cs

```diff
@@ -194,15 +241,107 @@ public override bool Execute ()
 				return true;
 
 			var codesignedFiles = new List<ITaskItem> ();
-			var resourcesToSign = Resources.Where (v => NeedsCodesign (v));
+			var resourcesToSign = Resources.Where (v => NeedsCodesign (v)).ToArray ();
+
+			// 1. Rewrite requests to sign executables inside frameworks to sign the framework itself
+			//    signing a framework and a file inside a framework is not *always* identical
+			//    on macOS apps {item.ItemSpec} can be a symlink to `Versions/Current/{item.ItemSpec}`
+			//    and `Current` also a symlink to `A`... and `_CodeSignature` will be found there
+			// 2. Resolve symlinks in the input.
+			// 3. Make sure we're working with full paths.
+			// All this makes it easier to sort and split the input files into buckets that can be codesigned together,
+			// while also not codesigning directories before files inside them.
+			foreach (var res in resourcesToSign) {
+				var path = res.ItemSpec;
+				var parent = Path.GetDirectoryName (path);
+
+				// so do not don't sign `A.framework/A`, sign `A.framework` which will always sign the *bundle*
+				if (Path.GetExtension (parent) == ".framework" && Path.GetFileName (path) == Path.GetFileNameWithoutExtension (parent))
+					path = parent;
+
+				path = PathUtils.ResolveSymbolicLinks (path);
+				path = Path.GetFullPath (path);
+
+				res.ItemSpec = path;
+			}
+
+			// first sort all the items by path length, longest path first.
+			resourcesToSign = resourcesToSign.OrderBy (v => v.ItemSpec.Length).Reverse ().ToArray ();
+
+			// Then we need to split the input into buckets, where everything in a bucket can be signed in parallel
+			// (i.e. no item in a bucket depends on any other item in the bucket being signed first).
+			// any such items must go into a different bucket. The bucket themselves are also sorted, where
+			// we have to sign the first bucket first, and so on.
+			// Since we've sorted by path length, we know that if we find a directory, we won't find any containing
+			// files from that directory later.
+			var buckets = new List<List<ITaskItem>> ();
+			for (var i = 0; i < resourcesToSign.Length; i++) {
```

Comment:
foreach and we get the need of the toArray?

Reply:
The order is important, and an IEnumerable isn't guaranteed to be sorted.